common/Resources/ti.internal/extensions/node/buffer.js

```diff
@@ -0,0 +1,1559 @@
+/**
+ * This implementation of Buffer uses a Ti.Buffer internally to back it.
+ * This is likley an order of magnitude slower than using a variant that extends Uint8Array!
+ * I think if we're not already wrapping a Ti.Buffer, it may be better to have two implementations
+ * and, like browserify, just extend Uint8Array for any Buffers we need to read/write a lot
+ * and then add a simple conversion method to turn it into a Ti.Buffer when needed.
+ *
+ * The Ti.Buffer impl has to go through the binding layer for reading/writing every byte.
+ * If we anticipate the Buffer staying on the JS side, I'm willing to bet that the Uint8Array
+ * the JS engine provides would be *way* faster.
+ *
+ * Also note that both Ti.Buffer and Node's Buffer were created before the JS engines had typed arrays
+ * (and Uint8Array in particular) as a means of encapsulating a byte array. We should consider accepting
+ * a Uint8Array in any of our APIs that take a Ti.Buffer and eventually deprecating/removing Ti.Buffer.
+ */
+
+// https://nodejs.org/api/buffer.html#buffer_buffers_and_character_encodings
+const TI_CODEC_MAP = new Map();
+TI_CODEC_MAP.set('utf-8', Ti.Codec.CHARSET_UTF8);
+TI_CODEC_MAP.set('utf8', Ti.Codec.CHARSET_UTF8);
+TI_CODEC_MAP.set('utf-16le', Ti.Codec.CHARSET_UTF16LE);
+TI_CODEC_MAP.set('utf16le', Ti.Codec.CHARSET_UTF16LE);
+TI_CODEC_MAP.set('ucs2', Ti.Codec.CHARSET_UTF16LE);
+TI_CODEC_MAP.set('ucs-2', Ti.Codec.CHARSET_UTF16LE);
+TI_CODEC_MAP.set('latin1', Ti.Codec.CHARSET_ISO_LATIN_1);
+TI_CODEC_MAP.set('binary', Ti.Codec.CHARSET_ISO_LATIN_1);
+TI_CODEC_MAP.set('ascii', Ti.Codec.CHARSET_ASCII);
+// We have no equivalents of base64 or hex, so we convert them internally here
+
+const VALID_ENCODINGS = [
+	'hex',
+	'utf8',
+	'utf-8',
+	'ascii',
+	'latin1',
+	'binary',
+	'base64',
+	'ucs2',
+	'ucs-2',
+	'utf16le',
+	'utf-16le'
+];
+
+// Used to cheat for read/writes of doubles
+const doubleArray = new Float64Array(1);
+const uint8DoubleArray = new Uint8Array(doubleArray.buffer);
+
+// Used to cheat to read/write floats
+const floatArray = new Float32Array(1);
+const uint8FloatArray = new Uint8Array(floatArray.buffer);
+
+class Buffer {
+	constructor(tiBuffer, start, length) {
+		this._tiBuffer = tiBuffer;
+		if (start === undefined) {
+			start = 0;
+		}
+		this.byteOffset = start;
+		if (length === undefined) {
+			this.length = tiBuffer.length - this.byteOffset;
+		} else {
+			this.length = length;
+		}
+		this._isBuffer = true;
+		// FIXME: Support .buffer property that holds an ArrayBuffer!
+	}
+
+	/**
+	 * 0 is returned if target is the same as buf
+	 *  is returned if target should come before buf when sorted.
+	 * -1 is returned if target should come after buf when sorted.
+	 * @param {Buffer} target Buffer to compare against
+	 * @param {integer} [targetStart=0] index to start in target
```

Comment:
Technically JSDoc can support more specialized numeric types; https://www.qooxdoo.org/current/pages/development/api_jsdoc_ref.html#handling-of-data-types

Basically I specify integer when possible because we really do expect an integer here and not a float/double - it gives more useful context/info.

Reply:
👍